sanitycheck: validate YAML files w/ schemas files

inakypg opened the pull request:
This introduces an schema-based YAML validation process when loading
any YAML file, before doing any operations on them. An exception will
be raised at SanityConfigParser() if the file fails to verify with the
given schema.

Schemas are defined for the platform files in board/*/*/*.yaml and for
the (sample|testcase).yaml files. The verification is done using the
pykwalify python library. If not installed, a warning is printed and
the verification schema is skipped. At some point, we might want to
force it being installed.

The verification library is made a separate module (scl.py) so it can
be easily imported by others.

Change-Id: Ib85fbdf6b5a412e846f5295d05725523b721b62f
Signed-off-by: Inaky Perez-Gonzalez <inaky.perez-gonzalez@intel.com>

dbkinder reviewed (dismissed):
Note that @nashif has a pending PR #799 that moves all the pip-installed packages into a requirements.txt file

dbkinder commented on doc/getting_started/installation_linux.rst:
```diff
@@ -55,16 +55,17 @@ Install the required packages in a Ubuntu host system with:
-
+	 python3-yaml dfu-util device-tree-compiler python3-pip
+   $ pip3 install --user pykwalify
+   
```
looks like there are some trailing blanks on this line

inakypg replied:
ah, missed amending that file after I corrected it -- ok, now it has the right one.

dbkinder commented:
Did you notice the odd test failures?  it's complaining about misspellings in scripts/spelling.txt  and issues with other files that aren't part of your patch?  something's kaput

inakypg commented:
@dbkinder , yeah, something is wrong -- it is complaining about things that are not in the patch -- whoever did that spell checking thing got it wrong. Not to mention the spell checker shouldn't dig on the spellcheck dbase that has the bad terms to check for ...

inakypg commented:
BTW, why is @dbkinder  made owner of these?

dbkinder commented:
I suspect I'm listed as THE code owner since I'm the only one in the CODEOWNERS file, as created by @galak

andrewboie commented on scripts/sanitycheck:
```diff
@@ -176,11 +176,17 @@ from collections import OrderedDict
 import yaml
 
+import logging
+log_format = "%(levelname)s %(name)s::%(module)s.%(funcName)s():%(lineno)d: %(message)s"
```
how is this used? i don't see any modifications for the verbose() or error() functions. move this to scl.py?

inakypg replied:
This sets the policy of how logging is to be reported; the modules/libraries log and the output is done based on what this policy says.

andrewboie commented on scripts/sanitycheck:
```diff
@@ -1236,31 +1241,40 @@ class TestSuite:
+                try:
+                    cp = SanityConfigParser(yaml_path, scl.yaml_tc_schema)
+                except RuntimeError as e:
+                    error("E: %s: can't load: %s" % (yaml_path, e))
```
the "E: " is redundant

inakypg replied:
conversation starter

The problem is error() is not printing any prefixes, so it is hard to tell what's a warning, what's an error, what's progress. I literally added that to be able to find the message while I was sprinkling errors to test.

I'd prefer error() and sibling prefix "Error:" "Warning:" or whatever (I prefer "[EWI]:" and then go from there, but I'd like your feedback on that first.

nashif reviewed (changes requested):
sanitycheck script already has some type and syntax checking, this needs to be removed to avoid duplication and confusion.

inakypg commented:
I don't think they overlap -- this is enforcing the YAML to be strictly well formatted; the get_section() check, which comes from the .ini format lack of strict formatting, does more than that (for example, it is splitting the arch_whitelist from a string into a list).

In order to get rid of get_section(), as it is now, the YAML would have to be modified to do things like the whitelist and such being a sequence vs a string -- so there would be a lot of extra work with possible side effects to no benefit.

I'd leave it as it is. It does well it's validation job and it does it well. My patch adds an extra level of checking appropriate to the format.

@andrewboie got any thoughts?

inakypg commented:
Last round of changes pushed - moving the loading of the Schemas to be done in sanity check itself rather than in scl.py.

inaky-intc committed:
sanitycheck: validate YAML files w/ schemas files

This introduces an schema-based YAML validation process when loading
any YAML file, before doing any operations on them. An exception will
be raised at SanityConfigParser() if the file fails to verify with the
given schema.

Schemas are defined for the platform files in board///*.yaml and for
the (sample|testcase).yaml files. The verification is done using the
pykwalify python library. If not installed, a warning is printed and
the verification schema is skipped. At some point, we might want to
force it being installed.

The verification library is made a separate module (scl.py) so it can
be easily imported by others.

Signed-off-by: Inaky Perez-Gonzalez <inaky.perez-gonzalez@intel.com>

dbkinder reviewed (dismissed):
+1 on docs, but needs merge conflict fixing
